apps/public/public.go

```diff
@@ -0,0 +1,251 @@
+		a.Account = account
+	}
+}
+
+// AcquireTokenSilent acquires a token from either the cache or using a refresh token.
+func (pca Client) AcquireTokenSilent(ctx context.Context, scopes []string, options ...AcquireTokenSilentOption) (AuthenticationResult, error) {
+	opts := AcquireTokenSilentOptions{}
+	for _, o := range options {
+		o(&opts)
+	}
+
+	silentParameters := client.AcquireTokenSilentParameters{
+		Scopes:      scopes,
+		Account:     opts.Account,
+		RequestType: accesstokens.RefreshTokenPublic,
+	}
+
+	return pca.Base.AcquireTokenSilent(ctx, silentParameters)
+}
+
+// AcquireTokenByUsernamePassword acquires a security token from the authority, via Username/Password Authentication.
+// NOTE: this flow is NOT recommended.
+func (pca Client) AcquireTokenByUsernamePassword(ctx context.Context, scopes []string, username string, password string) (AuthenticationResult, error) {
+	authParams := pca.AuthParams
+	authParams.Scopes = scopes
+	authParams.AuthorizationType = authority.AuthorizationTypeUsernamePassword
+	authParams.Username = username
+	authParams.Password = password
+
+	token, err := pca.Base.Token.UsernamePassword(ctx, authParams)
+	if err != nil {
+		return AuthenticationResult{}, nil
+	}
+	return pca.Base.AuthResultFromToken(ctx, authParams, token, true)
+}
+
+// DeviceCode provides the results of the device code flows first stage (containing the code)
+// that must be entered on the second device and provides a method to retrieve the AuthenticationResult
+// once that code has been entered and verified.
+type DeviceCode struct {
+	// Result holds the information about the device code (such as the code).
+	Result DeviceCodeResult
+
+	ctx        context.Context
+	authParams authority.AuthParams
+	client     Client
+	dc         requests.DeviceCode
+}
+
+type DeviceCodeResult = accesstokens.DeviceCodeResult
+
+// AuthenticationResult retreives the AuthenticationResult once the user enters the code
+// on the second device. Until then it blocks until the .AcquireTokenByDeviceCode() context
+// is cancelled or the token expires.
+func (d DeviceCode) AuthenticationResult() (AuthenticationResult, error) {
+	token, err := d.dc.Token()
+	if err != nil {
+		return AuthenticationResult{}, err
+	}
+	return d.client.AuthResultFromToken(d.ctx, d.authParams, token, true)
+}
+
+// AcquireTokenByDeviceCode acquires a security token from the authority, by acquiring a device code and using that to acquire the token.
+// Users need to create an AcquireTokenDeviceCodeParameters instance and pass it in.
+func (pca Client) AcquireTokenByDeviceCode(ctx context.Context, scopes []string) (DeviceCode, error) {
+	authParams := pca.AuthParams
+	authParams.Scopes = scopes
+	authParams.AuthorizationType = authority.AuthorizationTypeDeviceCode
+
+	dc, err := pca.Token.DeviceCode(ctx, authParams)
+	if err != nil {
+		return DeviceCode{}, err
+	}
+
+	return DeviceCode{Result: dc.Result, ctx: ctx, authParams: authParams, client: pca, dc: dc}, nil
```

Comment:
Not sure how I feel about the fact that we smuggle the provided context in the `DeviceCode` type.  It makes it unclear that `DeviceCode.AuthenticationResult()` makes IO requests.

Reply:
I don't think that a context object indicates IO, because it has a lot of uses beyond that.  But it did get me thinking....

I usually do this type of embedding when internally we will use the object as part of a single flow that has an overall timeout bounded by a context.  That is a good exception to the "no context as an attribute" rule.  

In this case, this probably qualifies as two operations that don't share a combined time.  I'll go back and look at his code and may remove the smuggling.  Update this in a bit.

Reply:
I've gone back and built a more traditional flow that should alleviate this concern.